flow/display_list_canvas_unittests.cc

```diff
@@ -45,25 +46,69 @@ constexpr SkRect RenderBounds =
     SkRect::MakeLTRB(RenderLeft, RenderTop, RenderRight, RenderBottom);
 
 class CanvasCompareTester {
- public:
+ private:
   // If a test is using any shadow operations then we cannot currently
   // record those in an SkCanvas and play it back into a DisplayList
   // because internally the operation gets encapsulated in a Skia
   // ShadowRec which is not exposed by their headers. For operations
   // that use shadows, we can perform a lot of tests, but not the tests
   // that require SkCanvas->DisplayList transfers.
   // See: https://bugs.chromium.org/p/skia/issues/detail?id=12125
-  static bool UsingShadows;
+  static bool TestingDrawShadows;
+  // The CPU renders nothing for drawVertices with a Blender.
+  // See: https://bugs.chromium.org/p/skia/issues/detail?id=12200
+  static bool TestingDrawVertices;
+  // The CPU renders nothing for drawAtlas with a Blender.
+  // See: https://bugs.chromium.org/p/skia/issues/detail?id=12199
+  static bool TestingDrawAtlas;
 
+ public:
   typedef const std::function<void(SkCanvas*, SkPaint&)> CvRenderer;
   typedef const std::function<void(DisplayListBuilder&)> DlRenderer;
 
+  // All of the tests should eventually use this method except for the
+  // tests that call |RenderNoAttributes| because they do not use the
+  // SkPaint object.
+  // But there are a couple of conditions beyond our control which require
+  // the use of one of the variant methods below (|RenderShadows|,
+  // |RenderVertices|, |RenderAtlas|).
   static void RenderAll(CvRenderer& cv_renderer, DlRenderer& dl_renderer) {
     RenderWithAttributes(cv_renderer, dl_renderer);
     RenderWithTransforms(cv_renderer, dl_renderer);
     RenderWithClips(cv_renderer, dl_renderer);
   }
 
+  // Used by the tests that render shadows to deal with a condition where
+  // we cannot recapture the shadow information from an SkCanvas stream
+  // due to the DrawShadowRec used by Skia is not properly exported.
+  // See: https://bugs.chromium.org/p/skia/issues/detail?id=12125
+  static void RenderShadows(CvRenderer& cv_renderer, DlRenderer& dl_renderer) {
+    TestingDrawShadows = true;
+    RenderNoAttributes(cv_renderer, dl_renderer);
+    TestingDrawShadows = false;
+  }
+
+  // Used by the tests that call drawVertices to avoid using an SkBlender
+  // during testing because the CPU renderer appears not to render anything.
+  // See: https://bugs.chromium.org/p/skia/issues/detail?id=12200
+  static void RenderVertices(CvRenderer& cv_renderer, DlRenderer& dl_renderer) {
+    TestingDrawVertices = true;
+    RenderAll(cv_renderer, dl_renderer);
+    TestingDrawVertices = false;
+  }
+
+  // Used by the tests that call drawAtlas to avoid using an SkBlender
+  // during testing because the CPU renderer appears not to render anything.
+  // See: https://bugs.chromium.org/p/skia/issues/detail?id=12199
```

Comment:
are we tracking this upstream bug anywhere so that we can fix these tests after that?

Reply:
It's low priority since we don't currently use the Blender property in Flutter, but I'll make sure it is added to "the lists".

Reply:
Added to the Skia issue tracker spreadsheet...